annubes/tasks/task.py

```diff
@@ -1,192 +1,205 @@
+            ntrials: Number of trials to generate.
+                Defaults to 20.
+            random_seed: Seed for numpy random number generator (rng). If an int is given, it will be used as the seed
+                for `np.random.default_rng`. If a dict is given, it must be in the form of a random state as given by
+                `rng.__getstate__()` (previous runs will have stored this value as `self.trials["random_state"]`).
+                Defaults to None (i.e. the initial state itself is random).
         """
-        # Attributes from the class constructor parameters
-        self.name = name
-        self.session_in = OrderedDict(session_in) if session_in else OrderedDict({"v": 0.5, "a": 0.5})
-        self.ordered = ordered
-        self.t_in = t_in
-        self.value_in = value_in if value_in else [0.8, 0.9, 1]
-        self.scaling = scaling
-        self.t_fixation = t_fixation
-        self.value_fixation = value_fixation
-        self.max_sequential = max_sequential
-        self.catch_prob = catch_prob
-        self.inter_trial = inter_trial
-        self.dt = dt
-        self.tau = tau
-        self.std_inp_noise = std_inp_noise
-        self.baseline_inp = baseline_inp
-        self.n_out = n_out
-        self.value_out = value_out if value_out else [0, 1]
+        self._ntrials = ntrials
 
-        # Derived attributes
-        self.trials = {}
-        self.trials["name"] = self.name
-        self.modalities = list(OrderedDict.fromkeys(char for string in session_in for char in string))
-        self.modality_idx = {m: i for i, m in enumerate(self.modalities)}
-        self.n_in = len(self.modalities) + 1  # +1 for start cue
-        self.value_in.sort()
-        self.imin = self.value_in[0]
-        self.imax = self.value_in[-1]
-        self.coeff = 0.6
-        if (self.value_fixation is not None) and self.scaling:
-            self.value_fixation = self._scale_input(self.value_fixation, self.coeff, self.imin, self.imax)
-        self.T = self.inter_trial + self.t_fixation + self.t_in
-        self.t = np.linspace(0, self.T, self.dt)
-        self.t = np.linspace(0, self.T, int((self.T + self.dt) / self.dt))
-        self.value_out.sort()
-        self.low_out = self.value_out[0]
-        self.high_out = self.value_out[1]
-
-        # Checks
-        # session_in
-        tolerance = 1 / 50
-        if not abs(sum(self.session_in.values()) - 1) < tolerance:
-            raise ValueError("The sum of the probabilities of `session_in` must be 1.")
-        # catch_prob
-        if not (catch_prob >= 0 and catch_prob < 1):
-            raise ValueError("`catch_prob` must be higher or equal to 0, or lower than 1.")
-
-    def _scale_input(
+        # Set random state
+        if isinstance(random_seed, dict):
+            np.random.default_rng().__setstate__(random_seed)
+        else:
+            rng = np.random.default_rng(random_seed)
```

Comment:
I still want to store the random state immediately after setting the rng, because it will have updated by the time we are creating the `trials` dictionary and it will not represent the state that was used in this had used to create it's randomness in the first place.